lib/Syntax/RawSyntax.cpp

```diff
@@ -67,10 +67,19 @@ static void dumpTokenKind(llvm::raw_ostream &OS, tok Kind) {
 
 } // end of anonymous namespace
 
+unsigned RawSyntax::NextFreeNodeId = 1;
+
 RawSyntax::RawSyntax(SyntaxKind Kind, ArrayRef<RC<RawSyntax>> Layout,
-                     SourcePresence Presence, bool ManualMemory) {
+                     SourcePresence Presence, bool ManualMemory,
+                     llvm::Optional<unsigned> NodeId) {
   assert(Kind != SyntaxKind::Token &&
          "'token' syntax node must be constructed with dedicated constructor");
+  if (NodeId.hasValue()) {
+    this->NodeId = NodeId.getValue();
+    NextFreeNodeId = std::max(this->NodeId + 1, NextFreeNodeId);
+  } else {
+    this->NodeId = NextFreeNodeId++;
+  }
```

Comment:
If the above comment (Receive `unsigned NodeId`) is true, you could factor out this: 
```cpp
static unsigned claimNodeId(unsigned NodeId) {
  if (!NodeId)
    return NextFreeNodeId++;
    
  NextFreeNodeId = std::max(NodeId + 1, NextFreeNodeId);
  return NodeId;
}
```
then call it from here and token constructor.
```cpp
this->NodeId = claimNodeId(NodeId);
```

Reply:
See above.